### PR TITLE
[Keyboard] Add Bubble 75

### PR DESCRIPTION
**Pull request by d-floe**
## Description

Bubble75 is a compact exploded F-Row 75% keyboard with Per-Key RGB. Includes Default and VIA keymaps.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/bubble75/hotswap/config.h [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+/* Mechanical locking support. Use KC_LCAP, KC_LNUM or KC_LSCR instead in keymap */
+#define LOCKING_SUPPORT_ENABLE
+
+/* Locking resynchronize hack */
+#define LOCKING_RESYNC_ENABLE
+
+/* key combination for command */
+#define IS_COMMAND() ( \
+    keyboard_report->mods == (MOD_BIT(KC_LSHIFT) | MOD_BIT(KC_RSHIFT)) \
+)
```
Comment (drashna) [inline keyboards/bubble75/hotswap/config.h:50]:
```suggetion
```

File: keyboards/bubble75/hotswap/config.h [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+#define RGBLIGHT_SAT_STEP 8
+#define RGBLIGHT_VAL_STEP 8
+#define RGBLIGHT_LAYERS
+//#define RGBLIGHT_LAYER_BLINK
+#define RGBLIGHT_LAYERS_OVERRIDE_RGB_OFF
+
+#define RGB_MATRIX_KEYPRESSES
+#define RGB_DISABLE_TIMEOUT 0
+#define RGB_DISABLE_AFTER_TIMEOUT 0
+#define RGB_DISABLE_WHEN_USB_SUSPENDED false
```
Comment (drashna) [inline keyboards/bubble75/hotswap/config.h:75]:
```suggestion
```

File: keyboards/bubble75/hotswap/hotswap.c [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "hotswap.h"
+#include "rgb_matrix_types.h"
```
Comment (drashna) [inline keyboards/bubble75/hotswap/hotswap.c:18]:
```suggestion
```

File: keyboards/bubble75/hotswap/hotswap.c [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    void suspend_power_down_kb(void) {
+        rgb_matrix_set_suspend_state(true);
+        suspend_power_down_user();
+}
+
+    void suspend_wakeup_init_kb(void) {
+        rgb_matrix_set_suspend_state(false);
+        suspend_wakeup_init_user();
+}
+
```
Comment (drashna) [inline keyboards/bubble75/hotswap/hotswap.c:64]:
```suggestion

File: keyboards/bubble75/hotswap/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,37 @@
+#include "hotswap.h"
```
Comment (drashna) [inline keyboards/bubble75/hotswap/keymaps/default/keymap.c:1]:
```suggestion
#include QMK_KEYBOARD_H
```

**Review by drashna — COMMENTED** [review]
It looks like you pulled this from kbfirmware.com directly. 

If so, you may want to run the config file through this: 
https://noroadsleft.github.io/kbf_qmk_converter/

And then overwrite the files here with what that outputs.

**Comment by d-floe** [comment]
> It looks like you pulled this from kbfirmware.com directly.
> 
> If so, you may want to run the config file through this: https://noroadsleft.github.io/kbf_qmk_converter/
> 
> And then overwrite the files here with what that outputs.
I don't have access to the original JSON made to build this Firmware, unfortunately. This is a little complicated for me.
I wrote up an info.json with KLE but I am unsure if it is sufficient
`{
    "keyboard_name": "bubble75", 
    "url": "", 
    "manufacturer": "phl",
    "maintainer": "velocifire", 
    "usb": {
        "vid": "0x4242",
        "pid": "0x5A4C",
    },
    "layouts": {
        "LAYOUT": {
            "layout": [{"x":0, "y":0}, 
                       {"x":1.5, "y":0}, 
                       {"x":2.5, "y":0}, 
                       {"x":3.5, "y":0}, 
                       {"x":4.5, "y":0}, 
                       {"x":6, "y":0}, 
                       {"x":7, "y":0}, 
                       {"x":8, "y":0}, 
                       {"x":9, "y":0}, 
                       {"x":10.5, "y":0}, 
                       {"x":11.5, "y":0}, 
                       {"x":12.5, "y":0}, 
                       {"x":13.5, "y":0}, 
                       {"x":15, "y":0}, 
                       {"x":0, "y":1.5}, 
                       {"x":1, "y":1.5}, 
                       {"x":2, "y":1.5}, 
                       {"x":3, "y":1.5}, 
                       {"x":4, "y":1.5}, 
                       {"x":5, "y":1.5}, 
                       {"x":6, "y":1.5}, 
                       {"x":7, "y":1.5}, 
                       {"x":8, "y":1.5}, 
                       {"x":9, "y":1.5}, 
                       {"x":10, "y":1.5}, 
                       {"x":11, "y":1.5}, 
                       {"x":12, "y":1.5}, 
                       {"x":13, "y":1.5, "w":2}, 
                       {"x":15, "y":1.5}, 
                       {"x":0, "y":2.5, "w":1.5}, 
                       {"x":1.5, "y":2.5}, 
                       {"x":2.5, "y":2.5}, 
                       {"x":3.5, "y":2.5}, 
                       {"x":4.5, "y":2.5}, 
                       {"x":5.5, "y":2.5}, 
                       {"x":6.5, "y":2.5}, 
                       {"x":7.5, "y":2.5}, 
                       {"x":8.5, "y":2.5}, 
                       {"x":9.5, "y":2.5}, 
                       {"x":10.5, "y":2.5}, 
                       {"x":11.5, "y":2.5}, 
                       {"x":12.5, "y":2.5}, 
                       {"x":13.5, "y":2.5, "w":1.5}, 
                       {"x":15, "y":2.5}, 
                       {"x":0, "y":3.5, "w":1.75}, 
                       {"x":1.75, "y":3.5}, 
                       {"x":2.75, "y":3.5}, 
                       {"x":3.75, "y":3.5}, 
                       {"x":4.75, "y":3.5}, 
                       {"x":5.75, "y":3.5}, 
                       {"x":6.75, "y":3.5}, 
                       {"x":7.75, "y":3.5}, 
                       {"x":8.75, "y":3.5}, 
                       {"x":9.75, "y":3.5}, 
                       {"x":10.75, "y":3.5}, 
                       {"x":11.75, "y":3.5}, 
                       {"x":12.75, "y":3.5, "w":2.25}, 
                       {"x":15, "y":3.5}, 
                       {"x":0, "y":4.5, "w":2.25}, 
                       {"x":2.25, "y":4.5}, 
                       {"x":3.25, "y":4.5}, 
                       {"x":4.25, "y":4.5}, 
                       {"x":5.25, "y":4.5}, 
                       {"x":6.25, "y":4.5}, 
                       {"x":7.25, "y":4.5}, 
                       {"x":8.25, "y":4.5}, 
                       {"x":9.25, "y":4.5}, 
                       {"x":10.25, "y":4.5}, 
                       {"x":11.25, "y":4.5}, 
                       {"x":12.25, "y":4.5, "w":1.75}, 
                       {"x":14, "y":4.5}, 
                       {"x":15, "y":4.5}, 
                       {"x":0, "y":5.5, "w":1.25}, 
                       {"x":1.25, "y":5.5, "w":1.25}, 
                       {"x":2.5, "y":5.5, "w":1.25}, 
                       {"x":3.75, "y":5.5, "w":6.25}, 
                       {"x":10, "y":5.5, "w":1.25}, 
                       {"x":11.25, "y":5.5, "w":1.25}, 
                       {"x":13, "y":5.5}, 
                       {"x":14, "y":5.5}, 
                       {"x":15, "y":5.5}]
        }
    }
}
`

File: keyboards/bubble75/hotswap/hotswap.c [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    4,4,4,4,4,4,4,4,4,4,4,4,  4,4,
+    4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,
+    4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,
+    4,4,4,4,4,4,4,4,4,4,4,4,4,  4,
+    4,4,4,4,4,4,4,4,4,4,4,4,  4,4,
+    4,4,4,    4,    4,4,  4,  4,4,
+  }
+};
+
+    void rgb_matrix_indicators_kb(void) {
```
Comment (drashna) [inline keyboards/bubble75/hotswap/hotswap.c:47]: Note to self: #18450

File: keyboards/bubble75/hotswap/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,37 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/bubble75/hotswap/keymaps/default/keymap.c:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/bubble75/hotswap/keymaps/via/keymap.c [hunk 8]
```diff
@@ -0,0 +1,37 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/bubble75/hotswap/keymaps/via/keymap.c:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/bubble75/hotswap/rules.mk [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+SLEEP_LED_ENABLE ?= no   # Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+NKRO_ENABLE ?= yes       # USB Nkey Rollover. See <https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work> if doesn't work.
+BACKLIGHT_ENABLE ?= no   # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE ?= no
+AUDIO_ENABLE ?= no       
+OLED_DRIVER_ENABLE ?= no 
+RGB_MATRIX_ENABLE = yes
+RGB_MATRIX_DRIVER = WS2812
+VIA_ENABLE = yes
+NO_USB_STARTUP_CHECK = yes
```
Comment (drashna) [inline keyboards/bubble75/hotswap/rules.mk:59]: Via should not be enabled at the keyboard level, only the keymap level. 

```suggestion
# Bootloader selection
BOOTLOADER = atmel-dfu

# Build Options
BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
MOUSEKEY_ENABLE = no        # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
NKRO_ENABLE = yes           # Enable N-Key Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
AUDIO_ENABLE = no           # Audio output

RGB_MATRIX_ENABLE = yes
RGB_MATRIX_DRIVER = WS2812
NO_USB_STARTUP_CHECK = yes
```

File: keyboards/bubble75/hotswap/keymaps/via/rules.mk [hunk 10]
```diff
@@ -0,0 +1 @@
+VIA_ENABLE ?=yes
```
Comment (drashna) [inline keyboards/bubble75/hotswap/keymaps/via/rules.mk:1]:
```suggestion
VIA_ENABLE = yes
```

File: keyboards/bubble75/hotswap/readme.md [hunk 11]
```diff
@@ -0,0 +1,22 @@
+# Bubble 75
+
+![image](https://user-images.githubusercontent.com/31980883/198159649-718a85f7-5d49-47b6-a46e-fa8c656d27c3.png)
```
Comment (drashna) [inline keyboards/bubble75/hotswap/readme.md:3]: Please rehost on imgur.

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by d-floe** [comment]
@drashna I made the changes requested, however, my Bubble75 no longer lets me change the RGB mode and doesn't get detected by VIA anymore. No errors while compiling.

**Comment by d-floe** [comment]
> @drashna I made the changes requested, however, my Bubble75 no longer lets me change the RGB mode and doesn't get detected by VIA anymore. No errors while compiling.
> 
> EDIT: My mistake, VIA seems to work fine. Still no ability to change RGB modes though. Im able to configure the lighting in VIA now, but the only lighting mode that works is solid color.

RGB has been fixed.

**Review by drashna — APPROVED** [review]

File: keyboards/bubble75/hotswap/hotswap.c [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        if (host_keyboard_led_state().caps_lock) {  // Capslock = WHITE
+            rgb_matrix_set_color(57, 255, 255, 255);
+    }
+}
```
Comment (drashna) [inline keyboards/bubble75/hotswap/hotswap.c:51]:
```suggestion
    return true;
}
```

**Comment by d-floe** [comment]
@drashna  Attempting to make these changes on my local copy results in an error when compiling
```Compiling: keyboards/bubble75/hotswap/hotswap.c                                                    keyboards/bubble75/hotswap/hotswap.c:47:6: error: conflicting types for 'rgb_matrix_indicators_kb'

 bool rgb_matrix_indicators_kb(void) {
      ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from quantum/quantum.h:36,
                 from keyboards/bubble75/hotswap/hotswap.h:19,
                 from keyboards/bubble75/hotswap/hotswap.c:17:
quantum/rgb_matrix/rgb_matrix.h:132:6: note: previous declaration of 'rgb_matrix_indicators_kb' was here
 void rgb_matrix_indicators_kb(void);
      ^~~~~~~~~~~~~~~~~~~~~~~~
keyboards/bubble75/hotswap/hotswap.c: In function 'rgb_matrix_indicators_kb':
keyboards/bubble75/hotswap/hotswap.c:48:9: error: invalid use of void expression
     if (!rgb_matrix_indicators_user()) {
         ^
keyboards/bubble75/hotswap/hotswap.c:55:1: error: control reaches end of non-void function [-Werror=return-type]
 }
 ^
cc1.exe: all warnings being treated as errors
 [ERRORS]
 |
 |
 |
make[1]: *** [builddefs/common_rules.mk:359: .build/obj_bubble75_hotswap_via/keyboards/bubble75/hotswap/hotswap.o] Error 1
Make finished with errors
make: *** [Makefile:414: bubble75/hotswap:via] Error 1

**Comment by d-floe** [comment]
> @drashna Attempting to make these changes on my local copy results in an error when compiling
> 
> ```
> Compiling: keyboards/bubble75/hotswap/hotswap.c                                                    In file included from quantum/rgb_matrix/rgb_matrix.h:23,
>                  from quantum/quantum.h:36,
>                  from keyboards/bubble75/hotswap/hotswap.h:19,
>                  from keyboards/bubble75/hotswap/hotswap.c:17:
> quantum/rgb_matrix/rgb_matrix_types.h:81:23: error: 'RGB_MATRIX_LED_COUNT' undeclared here (not in a function); did you mean 'RGB_MATRIX_ENABLE'?
>      led_point_t point[RGB_MATRIX_LED_COUNT];
>                        ^~~~~~~~~~~~~~~~~~~~
>                        RGB_MATRIX_ENABLE
> keyboards/bubble75/hotswap/hotswap.c: In function 'rgb_matrix_indicators_kb':
> keyboards/bubble75/hotswap/hotswap.c:55:1: error: control reaches end of non-void function [-Werror=return-type]
>  }
>  ^
> cc1.exe: all warnings being treated as errors
>  [ERRORS]
>  |
>  |
>  |
> make[1]: *** [builddefs/common_rules.mk:360: .build/obj_bubble75_hotswap_via/keyboards/bubble75/hotswap/hotswap.o] Error 1
> Make finished with errors
> make: *** [Makefile:414: bubble75/hotswap:via] Error 1
> ```
> 
> This is how I have formatted lines 47-55 of `hotswap.c`
> 
> ```
> bool rgb_matrix_indicators_kb(void) {
>     if (!rgb_matrix_indicators_user()) {
>         return false;
>     }
>     if (host_keyboard_led_state().caps_lock) {  // Capslock = WHITE
>         rgb_matrix_set_color(57, 255, 255, 255);
>       return true;
>     }
> }
> ```

Issue has been resolved with help from @waffle87 from the QMK discord.

File: keyboards/bubble75/hotswap/config.h [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+/* Set 0 if debouncing isn't needed */
+#define DEBOUNCE 5
+
+/* Mechanical locking support. Use KC_LCAP, KC_LNUM or KC_LSCR instead in keymap */
+#define LOCKING_SUPPORT_ENABLE
+
+/* Locking resynchronize hack */
+#define LOCKING_RESYNC_ENABLE
+
+/* prevent stuck modifiers */
```
Comment (drashna) [inline keyboards/bubble75/hotswap/config.h:41]:
```suggestion
```

File: keyboards/bubble75/hotswap/config.h [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+#define BOOTMAGIC_LITE_ROW 0
+#define BOOTMAGIC_LITE_COLUMN 0
+
+
```
Comment (drashna) [inline keyboards/bubble75/hotswap/config.h:46]: These are the default, so they don't need to be defined
```suggestion
```

File: keyboards/bubble75/hotswap/readme.md [hunk 15]
```diff
@@ -0,0 +1,22 @@
+# Bubble 75
+
+![image](https://i.imgur.com/9J01bYH.png)
```
Comment (drashna) [inline keyboards/bubble75/hotswap/readme.md:3]:
```suggestion
![image](https://i.imgur.com/9J01bYHh.png)
```

The additional `h` in the url/filename renders a smaller version of the image, that is still high res but better for github's markdown rendering

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/bubble75/hotswap/config.h [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+#define RGB_DISABLE_AFTER_TIMEOUT 0
+
+#define RGB_MATRIX_LED_PROCESS_LIMIT (RGB_MATRIX_LED_COUNT + 4) / 5
+#define RGB_MATRIX_LED_FLUSH_LIMIT 16
+#define RGB_MATRIX_MAXIMUM_BRIGHTNESS 180
+//#define RGB_MATRIX_STARTUP_MODE RGB_MATRIX_CYCLE_LEFT_RIGHT
+#define RGB_MATRIX_STARTUP_HUE 0
+#define RGB_MATRIX_STARTUP_SAT 255
+#define RGB_MATRIX_STARTUP_VAL RGB_MATRIX_MAXIMUM_BRIGHTNESS
+#define RGB_MATRIX_STARTUP_SPD 127
```
Comment (fauxpark) [inline keyboards/bubble75/hotswap/config.h:66]:
```suggestion
//#define RGB_MATRIX_DEFAULT_MODE RGB_MATRIX_CYCLE_LEFT_RIGHT
#define RGB_MATRIX_DEFAULT_HUE 0
#define RGB_MATRIX_DEFAULT_SAT 255
#define RGB_MATRIX_DEFAULT_VAL RGB_MATRIX_MAXIMUM_BRIGHTNESS
#define RGB_MATRIX_DEFAULT_SPD 127
```

File: keyboards/bubble75/hotswap/config.h [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+#define RGBLIGHT_HUE_STEP 8
+#define RGBLIGHT_SAT_STEP 8
+#define RGBLIGHT_VAL_STEP 8
+#define RGBLIGHT_LAYERS
+//#define RGBLIGHT_LAYER_BLINK
+#define RGBLIGHT_LAYERS_OVERRIDE_RGB_OFF
+
+#define RGB_MATRIX_KEYPRESSES
+#define RGB_DISABLE_TIMEOUT 0
+#define RGB_DISABLE_AFTER_TIMEOUT 0
```
Comment (fauxpark) [inline keyboards/bubble75/hotswap/config.h:57]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

**Comment by d-floe** [comment]
Hey there, any update on this? It's been radio silent for about a month. I don't mean to rush

File: keyboards/bubble75/hotswap/config.h [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+
+#pragma once
+
+#include "config_common.h"
+
+
+
+/* key matrix size */
+#define MATRIX_ROWS 6
+#define MATRIX_COLS 15
```
Comment (fauxpark) [inline keyboards/bubble75/hotswap/config.h:25]:
```suggestion
```

File: keyboards/bubble75/hotswap/rules.mk [hunk 19]
```diff
@@ -0,0 +1,19 @@
+# MCU name
+MCU = atmega32u4
+
+# Bootloader selection
+BOOTLOADER = atmel-dfu
+
```
Comment (fauxpark) [inline keyboards/bubble75/hotswap/rules.mk:6]:
```suggestion
```

File: keyboards/bubble75/hotswap/info.json [hunk 20]
```diff
@@ -0,0 +1,96 @@
+{
+    "keyboard_name": "bubble75", 
+    "url": "", 
+    "manufacturer": "phl",
+    "maintainer": "velocifire", 
+    "usb": {
+        "vid": "0x4242",
+        "pid": "0x5A4C",
+        "device_version": "0.0.1"
+    },
```
Comment (fauxpark) [inline keyboards/bubble75/hotswap/info.json:10]:
```suggestion
    },
    "processor": "atmega32u4",
    "bootloader": "atmel-dfu",
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Pull request merged**